File: benches/simplefrequency.rs
// Copyright 2021 Jonathan Manly.

// This file is part of rml.

// rml is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// rml is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.

// You should have received a copy of the GNU Lesser General Public License
// along with rml.  If not, see <https://www.gnu.org/licenses/>.

use criterion::{black_box, criterion_group, criterion_main, BenchmarkId, Criterion, Throughput};
use rml::math::norm;
use rml::preprocessing::text;
use rml::preprocessing::text::vectorizers::Ngrams;

const DATASIZES: [i32; 4] = [50, 100, 1000, 10000];

pub fn vectorizer_bench(c: &mut Criterion) {
    let mut group = c.benchmark_group("IMDB Vectorization");
    group.sample_size(50);

    let data = text::csv::parse_csv_with_labels::<String, String>(
        "./data/test_data/IMDB_Dataset.csv",
        true,
        text::csv::ClassPosition::Last,
    )
    .unwrap();
    let data = text::flatten(data.0);

    for i in DATASIZES {
        let mut fv = text::FrequencyVectorizerBuilder::new(i, true)
            .with_ngram_type(Ngrams::Both)
            .with_norm(norm::Norm::L2)
            .with_tfidf(true)
            .build();
        fv.gen_tokens(&data);
        println!("{:?}", fv.max_features);
        println!("{:?}", fv.get_tokens());
        group.throughput(Throughput::Elements(data.len() as u64));
        group.bench_with_input(
            BenchmarkId::new("IMDB FreqVec w/ Simple Tokenizer", i),
            &i,
            |b, _| b.iter(|| fv.vectorize(black_box(&data))),
        );
    }
}

criterion_group!(benches, vectorizer_bench);
criterion_main!(benches);

File: benches/simplefrequency.py
# Copyright 2021 Jonathan Manly.

# This file is part of rml.

# rml is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# rml is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.

# You should have received a copy of the GNU Lesser General Public License
# along with rml.  If not, see <https://www.gnu.org/licenses/>.

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer

from nltk.tokenize import TreebankWordTokenizer
import csv


features = []
with open("./data/test_data/IMDB_Dataset.csv", newline='', encoding='utf-8') as f:
    reader = csv.reader(f)
    for row in reader:
        if len(row) > 0:
            features.append(row[0])

# Normal


def vectorize(i):
    global features
    count_vect = CountVectorizer(max_features=i)
    tokenizer = TreebankWordTokenizer()
    count_vect.set_params(tokenizer=tokenizer.tokenize)
    count_vect.set_params(stop_words='english')

    count_vect.set_params(ngram_range=(1, 2))
    X_counts = count_vect.fit_transform(features)
    tfidf_transformer = TfidfTransformer()
    X_counts = tfidf_transformer.fit_transform(X_counts)


if __name__ == '__main__':
    import timeit
    import gc  # noqa:
    print(timeit.timeit("vectorize(50)",
                        setup="gc.enable()", number=1, globals=globals()))
    print(timeit.timeit("vectorize(100)",
                        setup="gc.enable()", number=1, globals=globals()))
    print(timeit.timeit("vectorize(1000)",
                        setup="gc.enable()", number=1, globals=globals()))
    print(timeit.timeit("vectorize(10000)",
                        setup="gc.enable()", number=1, globals=globals()))
